tests/DCPS/XTypes/XTypes.cpp

```diff
@@ -0,0 +1,670 @@
+
+  AppendableStructWithDependency as;
+  as.key = key_value;
+  as.additional_nested_struct.additional_field = NESTED_STRUCT_AF;
+  typed_dw->write(as, HANDLE_NIL);
+  if (verbose) {
+    ACE_DEBUG((LM_DEBUG, "writer: AppendableStructWithDependency\n"));
+  }
+}
+
+template<typename T>
+void get_topic(T ts, const DomainParticipant_var dp, const std::string& topic_name,
+  Topic_var& topic, const std::string& registered_type_name)
+{
+  ts->register_type(dp, registered_type_name.c_str());
+  CORBA::String_var type_name = (registered_type_name.empty() ? ts->get_type_name() : registered_type_name.c_str());
+  topic = dp->create_topic(topic_name.c_str(), type_name,
+    TOPIC_QOS_DEFAULT, 0, DEFAULT_STATUS_MASK);
+}
+
+
+bool check_inconsistent_topic_status(Topic_var topic)
+{
+  DDS::InconsistentTopicStatus status;
+  DDS::ReturnCode_t retcode;
+
+  retcode = topic->get_inconsistent_topic_status(status);
+  if (retcode != DDS::RETCODE_OK) {
+    ACE_ERROR((LM_ERROR, "ERROR: get_inconsistent_topic_status failed\n"));
+    return false;
+  } else if (status.total_count != (expect_to_match ? 0 : 1)) {
+    ACE_ERROR((LM_ERROR, "ERROR: inconsistent topic count is %d\n", status.total_count));
+    return false;
+  }
+  return true;
+}
+
+
+int ACE_TMAIN(int argc, ACE_TCHAR* argv[])
+{
+  DomainParticipantFactory_var dpf = TheParticipantFactoryWithArgs(argc, argv);
+  DomainParticipant_var dp = dpf->create_participant(23,
+    PARTICIPANT_QOS_DEFAULT, 0, DEFAULT_STATUS_MASK);
+
+  bool writer = false;
+  bool reader = false;
+  std::string type;
+  std::string registered_type_name = "";
+
+  for (int i = 1; i < argc; ++i) {
+    ACE_TString arg(argv[i]);
+    if (arg == ACE_TEXT("--verbose")) {
+      verbose = true;
+    } else if (arg == ACE_TEXT("--writer")) {
+      writer = true;
+    } else if (arg == ACE_TEXT("--reader")) {
+      reader = true;
+    } else if (arg == ACE_TEXT("--type")) {
+      if (i + 1 < argc) {
+        type = ACE_TEXT_ALWAYS_CHAR(argv[++i]);
+      } else {
+        ACE_ERROR((LM_ERROR, "ERROR: Invalid type argument"));
+        return 1;
+      }
+    } else if (arg == ACE_TEXT("--type_r")) {
+      if (i + 1 < argc) {
+        registered_type_name = ACE_TEXT_ALWAYS_CHAR(argv[++i]);
+      } else {
+        ACE_ERROR((LM_ERROR, "ERROR: Invalid registered type argument"));
+        return 1;
+      }
+    } else if (arg == ACE_TEXT("--key_val")) {
+      if (i + 1 < argc) {
+        key_value = ACE_OS::atoi(argv[++i]);
+      } else {
+        ACE_ERROR((LM_ERROR, "ERROR: Invalid key value argument"));
+        return 1;
+      }
+    } else if (arg == ACE_TEXT("--expect_to_fail")) {
+      expect_to_match = false;
+    } else {
+      ACE_ERROR((LM_ERROR, "ERROR: Invalid argument: %s\n", argv[i]));
+      return 1;
+    }
+  }
+
+  if (!(writer ^ reader)) {
+    ACE_ERROR((LM_ERROR, "ERROR: Invalid endpoint arguments\n"));
```

Comment:
Please use a more concrete error message to say user must specify either writer or reader but not both.